### PR TITLE
[auto parallel] complete elementwise ops spmd rules of LLaMa2 for eager semi  auto parallel 

### PR DESCRIPTION
**Pull request by cxxly**
### PR types

New features
### PR changes

Others
### Description

Pcard-73145

1. Overload elementwise grad spmdrules
 - ElementwiseUnaryGradInferSpmd(x, out, out_grad)
 - ElementwiseBinaryGradInferSpmd(x, y, out, out_grad)

2. Support bool dtype for sharding

3. Add yaml config and unittest  for follow ops, which are used in LLaMa2 model.

- multiply
- divide
- bitwise_and
- bitwise_not
- elementwise_pow
- equal
- exp
- rsqrt
- silu
- sin
- cos
- not_equal

4. Prompt rtol tolerance from 1e-5 to 1e-6

### CODE REVIEW

File: paddle/phi/infermeta/spmd_rules/elementwise.h [hunk 1]
```diff
@@ -30,6 +30,10 @@ SpmdInfo ElementwiseUnaryInferSpmdReverse(const DistMetaTensor& x,
 SpmdInfo ElementwiseUnaryGradInferSpmd(const DistMetaTensor& x,
                                        const DistMetaTensor& out_grad);
 
+SpmdInfo ElementwiseUnaryGradInferSpmd(const DistMetaTensor& x,
```
Comment (JZ-LIANG) [inline paddle/phi/infermeta/spmd_rules/elementwise.h:33]: 缺少单规则模块的单测，可以参考 test/auto_parallel/spmd_rules/test_elementwise_rule.py

File: paddle/phi/infermeta/spmd_rules/elementwise.h [hunk 2]
```diff
@@ -42,5 +46,11 @@ SpmdInfo ElementwiseBinaryGradInferSpmd(const DistMetaTensor& x,
                                         const DistMetaTensor& out_grad,
                                         int64_t axis = -1);
 
+SpmdInfo ElementwiseBinaryGradInferSpmd(const DistMetaTensor& x,
```
Comment (JZ-LIANG) [inline paddle/phi/infermeta/spmd_rules/elementwise.h:49]: 同上

File: paddle/phi/infermeta/spmd_rules/elementwise.h (hunk 1 above)
Comment (pkuzyc) [inline paddle/phi/infermeta/spmd_rules/elementwise.h:33]: 重载是为了适配 api 么？是不是可以加些注释说明一下

Comment (cxxly) [reply paddle/phi/infermeta/spmd_rules/elementwise.h:33]: 适配不同API签名，比如一元逐元素运算，部分算子反向只依赖输入，部分依赖输入输出，sin_grad(const Tensor& x, const Tensor& out_grad, Tensor* x_grad), silu_grad(const Tensor& x, const Tensor& out, const Tensor& out_grad, Tensor* x_grad)

**Review by JZ-LIANG — APPROVED** [review]

File: paddle/phi/infermeta/spmd_rules/elementwise.h (hunk 1 above)
Comment (cxxly) [reply paddle/phi/infermeta/spmd_rules/elementwise.h:33]: 新提一个PR补充

File: paddle/phi/infermeta/spmd_rules/elementwise.h (hunk 2 above)
Comment (cxxly) [reply paddle/phi/infermeta/spmd_rules/elementwise.h:49]: 新提交一个PR补充

**Review by XieYunshen — APPROVED** [review]
LGTM

**Review by zyfncg — APPROVED** [review]
LGTM for yaml file

**Pull request merged**